In [11]:
import numpy as np
import pandas as pd
from sklearn import tree

input_file = "C:/Users/Laiq Ahmed/Desktop/Project 4/breast-cancer-wisconsin.csv"
df = pd.read_csv(input_file, header = 0)

In [12]:
df.head()


Sample code number  Clump Thickness  Uniformity of Cell Size  \
0             1000025                5                        1   
1             1002945                5                        4   
2             1015425                3                        1   
3             1016277                6                        8   
4             1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

   Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0            1                3                1        1      2  
1           10                3                2        1      2  
2            2                3                1        1      2  
3            4                3                7        1      2  
4            1                3                1        1      2

In [13]:
df.head(50)

Sample code number  Clump Thickness  Uniformity of Cell Size  \
0              1000025                5                        1   
1              1002945                5                        4   
2              1015425                3                        1   
3              1016277                6                        8   
4              1017023                4                        1   
5              1017122                8                       10   
6              1018099                1                        1   
7              1018561                2                        1   
8              1033078                2                        1   
9              1033078                4                        2   
10             1035283                1                        1   
11             1036172                2                        1   
12             1041801                5                        3   
13             1043999                1                        1   
14             1044572                8                        7   
15             1047630                7                        4   
16             1048672                4                        1   
17             1049815                4                        1   
18             1050670               10                        7   
19             1050718                6                        1   
20             1054590                7                        3   
21             1054593               10                        5   
22             1056784                3                        1   
23             1057013                8                        4   
24             1059552                1                        1   
25             1065726                5                        2   
26             1066373                3                        2   
27             1066979                5                        1   
28             1067444                2                        1   
29             1070935                1                        1   
30             1070935                3                        1   
31             1071760                2                        1   
32             1072179               10                        7   
33             1074610                2                        1   
34             1075123                3                        1   
35             1079304                2                        1   
36             1080185               10                       10   
37             1081791                6                        2   
38             1084584                5                        4   
39             1091262                2                        5   
40             1096800                6                        6   
41             1099510               10                        4   
42             1100524                6                       10   
43             1102573                5                        6   
44             1103608               10                       10   
45             1103722                1                        1   
46             1105257                3                        7   
47             1105524                1                        1   
48             1106095                4                        1   
49             1106829                7                        8   

    Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                          1                  1                            2   
1                          4                  5                            7   
2                          1                  1                            2   
3                          8                  1                            3   
4                          1                  3                            2   
5                         10                  8        

In [14]:
df.dtypes

Sample code number             int64
Clump Thickness                int64
Uniformity of Cell Size        int64
Uniformity of Cell Shape       int64
Marginal Adhesion              int64
Single Epithelial Cell Size    int64
Bare Nuclei                    int64
Bland Chromatin                int64
Normal Nucleoli                int64
Mitoses                        int64
Class                          int64
dtype: object

In [52]:
#set the dataset into dependent and independent variables and into numpy array
X = df.iloc[:,1:9].values
Y = df.iloc[:,10].values
type(X)



numpy.ndarray

In [53]:
#split the data set into 75% training and 25% testing
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.25, random_state = 0 )

In [54]:
#scale the data (Feature Scalling)
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)


In [72]:
from sklearn.tree import DecisionTreeClassifier

In [73]:
dt = DecisionTreeClassifier(criterion = 'entropy')
dt.fit(X_train, Y_train)
dt.score(X_test, Y_test)

0.9428571428571428

In [60]:
from sklearn.model_selection import cross_val_score

In [62]:
from sklearn.ensemble import RandomForestClassifier

In [74]:
cross_val_score(DecisionTreeClassifier(criterion = 'entropy'),X,Y,cv=10)

array([0.88571429, 0.91428571, 0.95714286, 0.95714286, 0.98571429,
       0.97142857, 0.95714286, 0.97142857, 0.97142857, 0.94202899])

In [78]:
score1 = cross_val_score(DecisionTreeClassifier(criterion = 'entropy'),X,Y,cv=10)
np.average(score1)

0.939896480331263

As we can see above simple Decision tree classifier has a accuracy of 94.285% and with 10 cross folds it has a mean value of 93.98%, Hence Simple decision tree performs better in this case.  

In [65]:
rf = RandomForestClassifier(n_estimators=10)
rf.fit(X_train, Y_train)
rf.score(X_test, Y_test)

0.9771428571428571

In [68]:
cross_val_score(RandomForestClassifier(n_estimators=10),X,Y,cv=10)

array([0.9       , 0.94285714, 0.95714286, 0.91428571, 0.98571429,
       0.95714286, 0.95714286, 0.98571429, 0.97142857, 1.        ])

In [80]:
Score2 =cross_val_score(RandomForestClassifier(n_estimators=10),X,Y,cv=10)
np.average(Score2)

0.9471428571428572

The model Accuracy with Random Forest with 10 fold cross validation is 94.71% Hence it performs better than Decision tree.

In [93]:
from sklearn.neighbors import KNeighborsClassifier  
from sklearn import neighbors


In [178]:
clf = neighbors.KNeighborsClassifier(n_neighbors = 10)
clf.fit(X_train, Y_train)
clf.score(X_test, Y_test)

0.96

In [167]:
cross_val_score(KNeighborsClassifier(n_neighbors=10),X,Y,cv=10)

array([0.9       , 0.95714286, 0.95714286, 0.91428571, 1.        ,
       0.97142857, 0.98571429, 1.        , 0.98571429, 1.        ])

In [184]:
Score3 = cross_val_score(KNeighborsClassifier(n_neighbors=3),X,Y,cv=10)
np.average(Score3)

0.9713871635610767

The Accuracy Value for k = 10 is 96.714% if we reduce the K value to 3 we can get better results with the accuracy of 97.138%.

In [171]:
import sklearn
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
from sklearn import metrics
from sklearn.metrics import accuracy_score

In [140]:
GausNB = GaussianNB()
GausNB.fit(X_train, Y_train)
GausNB.score(X_test, Y_test)

0.96

In [141]:
cross_val_score(GaussianNB(),X,Y,cv=10)

array([0.92857143, 0.97142857, 0.97142857, 0.9       , 0.98571429,
       0.95714286, 0.97142857, 0.97142857, 0.98571429, 0.98550725])

In [142]:
Score4 = cross_val_score(GaussianNB(),X,Y,cv=10)

In [143]:
np.average(Score4)

0.9628364389233954

In [173]:
MultiNB = BernoulliNB()
MultiNB.fit(X_train, Y_train)
MultiNB.score(X_test, Y_test)

0.9428571428571428

In [175]:
cross_val_score(BernoulliNB(),X,Y,cv=10)

array([0.65714286, 0.65714286, 0.65714286, 0.65714286, 0.65714286,
       0.65714286, 0.65714286, 0.65714286, 0.64285714, 0.65217391])

In [176]:
score5 = cross_val_score(BernoulliNB(),X,Y,cv=10)

In [177]:
np.average(score5)

0.6552173913043479

The measure of accuracy for Gaussian (score4) with 10 cross validations is 96.2% for the model which is better if compare to multinomial(score5)